### PR TITLE
8312502: Mass migrate HotSpot attributes to the correct location

### PR DESCRIPTION
**Pull request by TheShermanTanker**
Someone had to do it, so I did. Moves attributes to the correct place as specified in the HotSpot Style Guide once and for all

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312502](https://bugs.openjdk.org/browse/JDK-8312502): Mass migrate HotSpot attributes to the correct location (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14969/head:pull/14969` \
`$ git checkout pull/14969`

Update a local copy of the PR: \
`$ git checkout pull/14969` \
`$ git pull https://git.openjdk.org/jdk.git pull/14969/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14969`

View PR using the GUI difftool: \
`$ git pr show -t 14969`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14969.diff">https://git.openjdk.org/jdk/pull/14969.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14969#issuecomment-1646424916)

### CODE REVIEW

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 1]
```diff
@@ -35,8 +35,8 @@
 
 #ifdef __GNUC__
 
-// __nop needs volatile so that compiler doesn't optimize it away
-#define NOP() asm volatile ("nop");
+// ISO C++ asm is always implicitly volatile
```
Comment (kimbarrett) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:38]: I can find no evidence for this claim, and it seems to me likely incorrect.  This is also way outside the
described scope of this PR.

File: src/hotspot/share/c1/c1_CFGPrinter.hpp [hunk 2]
```diff
@@ -62,7 +62,8 @@ class CFGPrinterOutput : public CHeapObj<mtCompiler> {
 
   void inc_indent();
   void dec_indent();
-  void print(const char* format, ...) ATTRIBUTE_PRINTF(2, 3);
+  ATTRIBUTE_PRINTF(2, 3)
+  void print(const char* format, ...);
```
Comment (kimbarrett) [inline src/hotspot/share/c1/c1_CFGPrinter.hpp:66]: This is an example where rearranging the attributes is out of character with usual practice.
And I think it makes it harder to read.

File: src/hotspot/share/compiler/compileLog.hpp [hunk 3]
```diff
@@ -69,7 +70,9 @@ class CompileLog : public xmlStream {
   // or reset, context string will be silently ignored
   stringStream* context()                        { return &_context; }
   void    clear_context()                        { context()->reset(); }
-  void      set_context(const char* format, ...) ATTRIBUTE_PRINTF(2, 3);
+
+  ATTRIBUTE_PRINTF(2, 3)
+  void      set_context(const char* format, ...);
```
Comment (kimbarrett) [inline src/hotspot/share/compiler/compileLog.hpp:75]: Whitespace between return type and function name is pretty pointless here.  Also above.

File: src/hotspot/share/utilities/globalDefinitions_gcc.hpp [hunk 4]
```diff
@@ -156,9 +156,9 @@ inline int g_isfinite(jdouble f)                 { return isfinite(f); }
 #define THREAD_LOCAL __thread
 
 // Inlining support
-#define NOINLINE     __attribute__ ((noinline))
-#define ALWAYSINLINE inline __attribute__ ((always_inline))
-#define ATTRIBUTE_FLATTEN __attribute__ ((flatten))
+#define NOINLINE     [[gnu::noinline]]
+#define ALWAYSINLINE [[gnu::always_inline]] inline
+#define ATTRIBUTE_FLATTEN [[gnu::flatten]]
```
Comment (kimbarrett) [inline src/hotspot/share/utilities/globalDefinitions_gcc.hpp:161]: This is way beyond the described scope of this PR.

File: src/hotspot/share/utilities/xmlstream.hpp [hunk 5]
```diff
@@ -109,29 +110,43 @@ class xmlStream : public outputStream {
...
+  void       va_done(const char* format, va_list ap);
 
   // write text (with quoting of special XML characters <>&'" etc.)
   outputStream* text() { return _text; }
-  void          text(const char* format, ...) ATTRIBUTE_PRINTF(2, 3);
-  void       va_text(const char* format, va_list ap) ATTRIBUTE_PRINTF(2, 0) {
+  ATTRIBUTE_PRINTF(2, 3)
+  void          text(const char* format, ...);
+  ATTRIBUTE_PRINTF(2, 0)
+  void       va_text(const char* format, va_list ap) {
```
Comment (kimbarrett) [inline src/hotspot/share/utilities/xmlstream.hpp:149]: This file is a particularly bad (to me) example of what happens without whitespace between a declaration
and the attributes for the next declaration.  I find this really hard to parse.  And the extra whitespace following
return types makes it even worse for me.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
Why?  What is the benefit from this that makes the resulting code churn
worthwhile?

We already discussed this kind of code churn a bit circa
https://github.com/openjdk/jdk/pull/11081#issuecomment-1313274792
and didn't like it then.  I don't see anything to change that.

The style guide only talks about the new C++ `[[attribute...]]` syntax, which
has a couple of valid locations. These are all gcc `__attribute__` and MSVC
`__declspec`, and are often located in places where the new syntax isn't
permitted. Moving the non-standard "attributes" around has the potential to
change semantics.  I don't know that it does, but this PR should contain
discussion and references to documentation showing it doesn't.

*If* it is to be done, there are some former one-liners that have been made
multi-line by moving an attribute macro, where there were multiple in a
cluster with no blank lines between them.  (Unwritten) HotSpot style only
elides whitespace between declarations when they are all one-liners.

I think I prefer preceding attributes to be on their own line, rather than on
the same line as the declaration, so I can mostly skip over them as I'm
reading the code. But that might be just me; I don't think that's been
discussed by the Group. It probably should have been brought up when
permitting attributes was added to the style guide, but it doesn't look like
that happened.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp (hunk 1 above)
Comment (TheShermanTanker) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:38]: Hi Kim, it's actually listed under https://gcc.gnu.org/onlinedocs/gcc/Basic-Asm.html as follows:

> Qualifiers
volatile
The optional volatile qualifier has no effect. All basic asm blocks are implicitly volatile.

I'll take this outside of this PR though

File: src/hotspot/share/utilities/globalDefinitions_gcc.hpp (hunk 4 above)
Comment (TheShermanTanker) [reply src/hotspot/share/utilities/globalDefinitions_gcc.hpp:161]: I changed these to the standard attributes so the compilers would concretely enforce the checks of which areas the attributes appertained to, as opposed to the regular __attribute__ syntax which don't perform such checks

**Comment by theRealAph** [comment]
> Why? What is the benefit from this that makes the resulting code churn worthwhile?
> 
> We already discussed this kind of code churn a bit circa [#11081 (comment)](https://github.com/openjdk/jdk/pull/11081#issuecomment-1313274792) and didn't like it then. I don't see anything to change that.

I agree. Such changes don't much help maintainers, and speaking as the lead of both the 8u and 11u projects, really won't help backporting. I'd say no.

File: src/hotspot/share/c1/c1_CFGPrinter.hpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/c1/c1_CFGPrinter.hpp:66]: I agree with Kim, I do not like this style of using a new line for the attribute. I also prefer to see these attributes in their original location where I can generally ignore them while reading the code.

File: src/hotspot/share/utilities/globalDefinitions_gcc.hpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/share/utilities/globalDefinitions_gcc.hpp:161]: Again I agree with Kim, this is not simply moving an attribute to a new location. It is a change that has to be validated.

File: src/hotspot/share/utilities/xmlstream.hpp (hunk 5 above)
Comment (dholmes-ora) [reply src/hotspot/share/utilities/xmlstream.hpp:149]: +1

**Comment by dholmes-ora** [comment]
> Someone had to do it, so I did. 

Why did someone _have_ to do it? Is it incorrect?

Unless there is some semantic significance to this then it is just unnecessary churn and I really don't like the chosen style. Sorry.

**Comment by kimbarrett** [comment]
See also:
https://openjdk.org/guide/#things-to-consider-before-proposing-changes-to-openjdk-code
This change looks like a case of pure "Modernizing", which isn't looked on
particularly favorably for its own sake. There generally needs to be some
additional benefits.

**Comment by kimbarrett** [comment]
> Hmm... I was doing this so that what attribute specifiers applied to would be clearer since it's helpful to have the ability to make them apply to different aspects of a method as per C++11. Could I split this up and discard some of the areas in which the changes are less agreeable? (as advised by someone who reached out to me privately)

That can be done when other attributes are being added to a declaration. For
example, various error reporting functions had the `[[noreturn]]` attribute
added by JDK-8303805. Many of them also had a trailing ATTRIBUTE_PRINTF, which
was moved to the front at the same time. There were also one or two functions
there whose trailing ATTRIBUTE_PRINTF was moved to the front even though not
noreturn, for local consistency (e.g. `warning`).  (So yes, sometimes it's
okay to normalize some outliers.  That doesn't seem like what's going on in
this PR though.)

There currently aren't other standard attributes to be dealt with.  That may
change as we adopt newer versions of the standard.  But even so, the scope of
the adoption changes is going to be limited.

**Comment by theRealAph** [comment]
> Could I split this up and discard some of the areas in which the changes are less agreeable? (as advised by someone who reached out to me privately)

I'm a little surprised that you're persisting with this. 

Kim said:
> https://openjdk.org/guide/#things-to-consider-before-proposing-changes-to-openjdk-code
> This change looks like a case of pure "Modernizing", which isn't looked on
> particularly favorably for its own sake. There generally needs to be some
> additional benefits.
